pyquil/api/_qam.py

```diff
@@ -13,133 +13,67 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 ##############################################################################
-import warnings
 from abc import ABC, abstractmethod
-from collections import defaultdict
-from typing import Dict, Sequence, Union, Optional
+from dataclasses import dataclass, field
+from typing import Dict, Generic, Optional, TypeVar
 
 import numpy as np
-from rpcq.messages import ParameterAref
-
 from pyquil.api._abstract_compiler import QuantumExecutable
-from pyquil.api._error_reporting import _record_call
 from pyquil.experiment._main import Experiment
 
 
 class QAMError(RuntimeError):
     pass
 
 
-class QAM(ABC):
-    """
-    The platonic ideal of this class is as a generic interface describing how a classical computer
-    interacts with a live quantum computer.  Eventually, it will turn into a thin layer over the
-    QPU and QVM's "QPI" interfaces.
+ExecuteResponse = TypeVar("ExecuteResponse")
```

Comment:
I think we should call this a "job" rather than a "response" to convey the intention, especially since that's how it's used in the migration docs.

Reply:
Applies mostly to the actual type arguments in `QPU` and `QVM` subclasses. For the type variable, I think the pattern is to use something short, like `TypeVar("T")` (see: https://docs.python.org/3/library/typing.html#generics)

Reply:
It's only a job for QAM though - for QVM and PyQVM it's not a job.

Reply:
Good point. 👍 You may still want to name this T to follow convention, but no big deal

Reply:
will do